src/snackbar/SnackbarManager.tsx

```diff
@@ -0,0 +1,111 @@
+import { Snackbar, withTheme, WithTheme } from "@material-ui/core";
+import { arrayEquals } from "@xenit/finder-utils";
+import * as React from "react";
+import SnackbarNotification, { SnackbarNotification_Variant_t } from "./SnackbarNotification";
+import SnackbarTraceDialog from "./SnackbarTraceDialog";
+
+export type SnackbarManager_Message_t = {
+    type: SnackbarNotification_Variant_t,
+    message: string,
+    expires?: boolean,
+    trace?: string,
+};
+
+type SnackbarManager_Props_t<Message extends SnackbarManager_Message_t> = {
+    messages: readonly Message[],
+    onMessageExpired?: (message: Message) => void,
+};
+
+function findNextMessage<T extends SnackbarManager_Message_t>(messages: readonly T[], closedMessages: readonly T[]): { next: T | null, closedMessages: T[] } {
+    let next = null;
+    const newClosedMessages: T[] = [];
+    for (const message of messages) {
+        if (closedMessages.indexOf(message) === -1) {
+            if (next === null) {
+                next = message;
+            }
+        } else {
+            newClosedMessages.push(message);
+        }
+    }
+    return { next, closedMessages: newClosedMessages };
+
+}
+
+export default function SnackbarManager<T extends SnackbarManager_Message_t>({ messages, onMessageExpired }: SnackbarManager_Props_t<T>) {
+    let [closedMessages, setClosedMessages] = React.useState([] as readonly T[]);
+
+    const { next: nextMessage, closedMessages: newClosedMessages } = findNextMessage(messages, closedMessages);
+    if (!arrayEquals(newClosedMessages, closedMessages.slice())) {
+        setClosedMessages(newClosedMessages);
+    }
+    closedMessages = newClosedMessages;
```

Comment:
Why this assignment if the arrays have the same contents now anyway?

Reply:
Maybe I should add a comment there why I did this.

React state is not updated immediately, the value remains the same for this render.
When the `onClose` method below is called in the same render, I want the correct state to be set, not one where the previously closed message is now no longer in the closed array.

https://github.com/xenit-eu/finder-ui/blob/574156310e37b9b16a134fc88963fafc683d63f4/src/snackbar/SnackbarManager.tsx#L46-L53

However, I am wrong here, because state updates are async, as documented on https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous, I should have used `setClosedMessages(closedMessages => closedMessages.concat([nextMessage]))` instead.

Reply:
Ok, that makes sense then